File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        Set<Character> opens = Set.of('(', '{', '[');
        for(int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (opens.contains(c)) {
                stack.push(c);
            } else {
                if(stack.isEmpty()) return false;

                char top = stack.peek();

                if((c == ')' && top == '(') ||
                   (c == ']' && top == '[') ||
                   (c == '}' && top == '{')) {
                    stack.pop();
                   } else {
                    return false;
                   }
            }
        }

        return stack.isEmpty();
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mappings = {")" : "(", "}" : "{", "]" : "["}
        for c in s:
            if c in mappings:
                if stack and mappings[c] == stack[-1]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)
        return True if not stack else False
            
        